### PR TITLE
Implement Order handling

### PR DESCRIPTION
**Pull request by cbj4074**
More than anything, this PR is a rough-cut.

The first unusual aspect of it is that the `Order` model ID is a UUID (rather than an auto-incremented sequential value). The rationale is that site visitors are not logged-in when adding Items to their Orders, and in the interest of preventing various order hijacking attacks, there must exist a means by which to associate a user-agent with a specific `Order` in a way that leverages a crypto-grade random token of some sort. In other words, site visitors must be able to shop in a "guest" capacity.

To elaborate, imagine that an unauthenticated visitor wishes to add an Item to an Order. The API call that returns the Order ID shouldn't return a predictable, auto-incremented Order ID that a malicious actor could "guess" easily. In essence, the challenge here is to provide a means by which to "authenticate" changes to a given order, absent some other form of authentication (such as a user being logged-in).

Similarly, operations performed against individual `Order Item`s shouldn't be permitted given only an `Order Item` ID (these IDs are sequential and therefore predictable). Rather, it's necessary to provide a parent order ID (which, again, is a randomly-generated UUID, using a crypto-grade PRNG) to act on a given `Order Item`.

I'll annotate specific areas of the code that I think could be improved in some way, or that regarding which I'd like to solicit additional feedback.

### CODE REVIEW

File: app/Repositories/OrderItemRepository.php [hunk 1]
```diff
@@ -39,6 +40,29 @@ public function class()
...
+
+    public function create(array $data)
+    {
+        try {
+            return $this->model()->create($data);
+        } catch (QueryException $e) {
+            // If Integrity Constraint violation, the same item has already
+            // been added to the Order, and we should simply ignore the failure.
+
+            if ($e->getCode() !== '23000') {
```
Comment (cbj4074) [inline app/Repositories/OrderItemRepository.php]: I don't love that this SQL error code is hard-coded and specific to MySQL/MariaDB. I suppose that we could abstract the logic further and add a method that works for every database engine. More details regarding a cross-engine approach: https://stackoverflow.com/a/26371673/1772379

As to the rationale, this unique key is intended to prevent the customer from adding a given item to the order more than once. Given that all items are digital, my position is that there is no value in ordering multiple copies of a given item (song or album).

We allowed for this in the legacy implementation, so that customers could pay artists more than requested for a given item (by increasing the quantity of a given song), and while it was used on occasion, the feature seems ripe for abuse in a high-visibility deployment.

Specifically, credit card fraudsters might be keen to order 99 copies of a song, for example, hoping to channel the payments to an artist under their control before being detected. At a minimum, preventing quantities greater than one will make that type of abuse more cumbersome for bad actors.

File: app/Services/ApiRoute.php [hunk 2]
```diff
@@ -115,7 +114,8 @@ public function addRoute($uri, $controllerMethod, $httpMethod = 'get', $name = n
                     $controllerMethod,
                     $httpMethod
                 ) {
-                    Route::$httpMethod($uri, "$controller@$controllerMethod")->name("$prefix.$name");
+                    Route::$httpMethod($uri, "$controller@$controllerMethod")
+                        ->name("$prefix." . snake_case($controllerMethod));
```
Comment (cbj4074) [inline app/Services/ApiRoute.php]: I made this change to fix an inconsistency in the generated name vs. the pattern that we've used historically. I can't recall the precise details, but this didn't break any tests and now the names are as I expect.

File: app/UuidModel.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        parent::__construct($attributes);
+    }
+
+    public static function boot()
+    {
+        parent::boot();
+
+        self::creating(function ($model) {
+            $model->{$model->getKeyName()} = Uuid::uuid4()->toString();
```
Comment (cbj4074) [inline app/UuidModel.php]: In contrast to other UUID versions, UUID4 is designed specifically to be random, which is why it's used here. The intention herein is for this value to be very difficult to guess, so as to prevent malicious user-agents from hijacking other users' orders.

File: app/Http/Controllers/Api/OrderController.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        return UpdateOrder::class;
+    }
+
+    /**
+     * Create an Order, which requires adding one or more Items to the Order.
+     *
+     * @param Request $request
+     * @return Response
+     */
+    public function storeOrder(StoreOrder $request)
```
[the remark this answers is not in the document]
Comment (poppabear8883) [reply app/Http/Controllers/Api/OrderController.php]: 1.) The method you propose to solve the unauthorized manipulation seems to be acceptable IMO, However I'm not so sure that I am the most qualified in this particular area to say if its the BEST method.

2.) I'm confused on the TODO issue tbh ... Not sure what the requirement vs complication is as its not entirely clear.

3.) Suggestion on Request Class naming conventions appending `Request` ... ie: `StoreOrderRequest` . Whats your take on this ?

File: app/Repositories/OrderItemRepository.php (hunk 1 above)
Comment (poppabear8883) [reply app/Repositories/OrderItemRepository.php]: So why not simply do an OR conditional here and hard code the value per supported (or possible future supported) database responses ? Is the `QueryException` here driver driven ? I assume it is where `->getCode();` will return the code on any given driver used. Which means when just need to hard code every possible code that it could be and call it a day as it will not ever likely change in the future anyways.

File: app/Services/ApiRoute.php (hunk 2 above)
Comment (poppabear8883) [reply app/Services/ApiRoute.php]: By changing this is `$name` now unused variable ? `line 96` here should probably be what you need to change to apply the `snake_case()` method. Correct me if I am wrong here ?

File: app/UuidModel.php (hunk 3 above)
Comment (poppabear8883) [reply app/UuidModel.php]: You are definitely more qualified to handle anything related to UUID as I'm not familiar with it, so LGTM :laughing:

File: app/Http/Controllers/Api/OrderController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbj4074) [reply app/Http/Controllers/Api/OrderController.php]: 1.) Very good. I've spent enough time reasoning through it that I'm comfortable running with it if and until a better approach emerges.

2.) Sorry, I should have been more explicit regarding the TODO.

Fundamentally, the problem I encountered is that the pattern we're employing currently prevents us from type-hinting the Request object on the various controller methods.

If you look at the methods I added that don't conform to our existing pattern, `storeOrder()`, `addItems()` and `removeItems()`, they all deviate in that they type-hint a Request object. Basically, we need access to the Request object to be able to retrieve and manipulate the POST inputs in a way that is more complex than has been required elsewhere in the code-base to date.

Basically, I created the `storeOrder(StoreOrder $request)` method to use in place of the "built-in" `storeRequest()` method, because I need access to the Request object (via the `$request` variable as seen in the method signature). The problem is that we can't type-hint like this on the `storeRequest()` method, e.g., `storeRequest(StoreOrder $request)`, because doing so results in `Declaration must be compatible with ApiController->storeRequest()`.

In short, if you can find a way to achieve what I've done in those three methods while adhering to the current pattern, I'd be very curious to see how you do it.

3.) I like it. I'm happy to rename them.

File: app/Repositories/OrderItemRepository.php (hunk 1 above)
Comment (cbj4074) [reply app/Repositories/OrderItemRepository.php]: Indeed; let's keep it simple. Yes, exactly, regarding `getCode()`. I'll add the codes for the other drivers that Laravel supports and call it a wrap.

File: app/Services/ApiRoute.php (hunk 2 above)
Comment (cbj4074) [reply app/Services/ApiRoute.php]: You're absolutely correct; my mistake. I think what we actually want is simply to remove the former line 96 (as in this new commit), but at the same time change the new line 96 to this:

```php
$name = is_null($name) ? snake_case($controllerMethod) : $name;
```

That gives me the results I expect, all the tests pass, and the `$name` is no longer ignored.

File: app/Services/ApiRoute.php [hunk 5]
```diff
@@ -93,8 +93,7 @@ public function addRoute($uri, $controllerMethod, $httpMethod = 'get', $name = n
         $controller = $this->controller;
         $prefix = $this->prefix;
         $httpMethod = strtolower($httpMethod);
-        $controllerMethod = snake_case($controllerMethod);
-        $name = is_null($name) ? strtolower($controllerMethod) : strtolower($name);
+        $name = is_null($name) ? snake_case($controllerMethod) : $name;
```
Comment (cbj4074) [inline app/Services/ApiRoute.php]: @poppabear8883 Does this look better?

Comment (poppabear8883) [reply app/Services/ApiRoute.php]: LGTM

File: app/Http/Controllers/Api/OrderController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (poppabear8883) [reply app/Http/Controllers/Api/OrderController.php]: Regarding the TODO, if you have a working solution right this minute i say lets move forward, but i would like to make note to revisit this once i'm back at it. I think there could be better approach to this but am not really sure what that is atm. Instead of tying this PR up lets get it merged in.

Could you please open a issue regarding this topic so that it doesn't get lost ?

[the remark this answers is not in the document]
Comment (cbj4074) [reply app/Http/Controllers/Api/OrderController.php]: I concur! Yes, I'll create an Issue about it right now, so we don't forget.
